### PR TITLE
Add certificate pinning

### PR DESCRIPTION
**Pull request by ffflorian**

### CODE REVIEW

File: electron/main.js [hunk 1]
```diff
@@ -168,6 +169,20 @@ function showMainWindow() {
     main.setBounds(init.restore('bounds', main.getBounds()));
   }
 
+  main.webContents.session.setCertificateVerifyProc((hostname = '', certificate = {}, cb) => {
+    const { data: cert = '' } = certificate;
```
Comment (lipis) [inline electron/main.js]: don't add spaces after and before the `{}`

```
{data: cert = ''}
```

File: electron/js/certutils.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+const MAIN_CERT = 'ff4880f09cc1e2d0edc0d07fdcdc987c2b5f4d9c';
+const strip = (url) => url.replace(/https:|[\/]+/g, '');
+const pins = [
+  { url: strip(config.PROD_URL),      keys: [ 'c85f71715e51593828e37cd8450501f3fa5fa4a2',
+                                              MAIN_CERT ] },
+  { url: 'wire.com',                  keys: [ MAIN_CERT ] },
+  { url: 'www.wire.com',              keys: [ MAIN_CERT ] },
+  { url: 'prod-nginz-https.wire.com', keys: [ MAIN_CERT ] },
+  { url: 'prod-nginz-ssl.wire.com',   keys: [ MAIN_CERT ] },
+  { url: 'prod-assets.wire.com',      keys: [ MAIN_CERT ] }
```
Comment (lipis) [inline electron/js/certutils.js]: remove spaces after/before `{}`: `{url: 'prod-assets.wire.com',      keys: [ MAIN_CERT ]}`

File: electron/js/certutils.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+];
+
+const str2ab = (str) => {
+  const buf = new ArrayBuffer(str.length);
+  const bufView = new Uint8Array(buf);
+  for (let i = 0, strLen = str.length; i < strLen; i++) {
```
Comment (lipis) [inline electron/js/certutils.js]: `i` -> `index`

File: electron/js/certutils.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    bufView[i] = str.charCodeAt(i);
+  }
+  return buf;
+};
+
+const pem2cert = (pem) => {
+  const strippedPem = new String(pem).replace(/(-----(BEGIN|END) CERTIFICATE-----|\n)/g, '').trim();
+  const raw = new Buffer(strippedPem, 'base64').toString('binary');
+  const asn1 = asn1js.fromBER(str2ab(raw)).result;
+  return new pkijs.Certificate({ schema: asn1 });
```
Comment (lipis) [inline electron/js/certutils.js]: spaces

File: package.json [hunk 5]
```diff
@@ -47,6 +47,7 @@
     "grunt-git": "1.0.3",
     "grunt-github-changes": "0.1.0",
     "grunt-shell": "2.1.0",
-    "load-grunt-tasks": "3.5.2"
-  }
+    "load-grunt-tasks": "3.5.2",
+    "pkijs": "^2.0.48"
```
Comment (lipis) [inline package.json]: remove `^`

File: electron/main.js [hunk 6]
```diff
@@ -168,6 +169,20 @@ function showMainWindow() {
...
+
+    if (certutils.hostnameShouldBePinned(hostname) === true) {
+      if (certutils.verifyPinning(hostname, cert) === true) {
```
Comment (lipis) [inline electron/main.js]: skip `=== true`

File: electron/js/certutils.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  verifyPinning (hostname, cert) {
+    const certKey = getPublicKeyHash(cert);
+    let foundKey = false;
+
+    for (let pin of pins) {
+      if (pin.url === hostname) {
+        for (let key of pin.keys) {
+          if (key.toLowerCase().trim() === certKey.toLowerCase().trim()) {
+            foundKey = true;
```
Comment (lipis) [inline electron/js/certutils.js]:
```
return true
```

File: electron/js/certutils.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            break;
+          }
+        }
+        break;
+      }
+    }
+    return foundKey;
```
Comment (lipis) [inline electron/js/certutils.js]: return `false`

File: electron/js/certutils.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      if (pin.url.toLowerCase().trim() === hostname.toLowerCase().trim()) {
+        foundPin = true;
+        break;
+      }
+    }
+    return foundPin;
+  },
+
+  verifyPinning (hostname, cert) {
+    const certKey = getPublicKeyHash(cert);
```
Comment (lipis) [inline electron/js/certutils.js]:
```
const certKey = getPublicKeyHash(cert).toLowerCase().trim()
```

File: electron/main.js (hunk 6 above)
Comment (ffflorian) [inline electron/main.js]: Why? I want to check it exactly, not truthy.

File: electron/js/certutils.js (hunk 2 above)
Comment (ffflorian) [inline electron/js/certutils.js]: But not after/before `[]`?

File: electron/js/certutils.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  const bufView = new Uint8Array(buf);
+  for (let index = 0, strLen = str.length; index < strLen; index++) {
+    bufView[index] = str.charCodeAt(index);
+  }
+  return buf;
+};
+
+const pem2cert = (pem) => {
+  const strippedPem = new String(pem).replace(/(-----(BEGIN|END) CERTIFICATE-----|\n)/g, '').trim();
+  const raw = new Buffer(strippedPem, 'base64').toString('binary');
```
Comment (herrmannplatz) [inline electron/js/certutils.js]: would be nice to use `Buffer.from`https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_string_encoding

File: electron/js/certutils.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  const asn1 = asn1js.fromBER(str2ab(raw)).result;
+  return new pkijs.Certificate({schema: asn1});
```
Comment (herrmannplatz) [inline electron/js/certutils.js]: just pass the buffer to asn1js.fromBER

File: electron/main.js (hunk 6 above)
Comment (lipis) [inline electron/main.js]: what else can it be?! since we are in control of these functions..

File: electron/main.js [hunk 12]
```diff
@@ -168,6 +169,20 @@ function showMainWindow() {
     main.setBounds(init.restore('bounds', main.getBounds()));
   }
 
+  main.webContents.session.setCertificateVerifyProc((hostname = '', certificate = {}, cb) => {
+    const {data: cert = ''} = certificate;
+
+    if (certutils.hostnameShouldBePinned(hostname) === true) {
+      if (certutils.verifyPinning(hostname, cert) === true) {
```
Comment (lipis) [inline electron/main.js]: we are in control of these functions.. so it's either `true` or `false` so no need to `=== true`

**Pull request merged**